### PR TITLE
fix: Resolve hostname with IPv4 first

### LINKED ISSUE #19931: 🐛 Bug Report: Fetch error Node 18.x k8s plugin LocalKubectlProxyClusterLocator
### 📜 Description

Running the K8s plugin in Node 18.x with `LocalKubectlProxy` gives a fetch error `ECONNREFUSED ::1:8001`.

We believe this being an issue with how the `LocalKubectlProxyClusterLocator` declares the url of the proxy with localhost, then `node-fetch` using the native Node 18.x Fetch, and that fetch not liking the `localhost` part, so maybe changing that to loopback IP `127.0.0.1` could be the fix.

https://github.com/backstage/backstage/blob/69abd07a44555b3a09c2cd3d6f5dda59c14de3f5/plugins/kubernetes-backend/src/cluster-locator/LocalKubectlProxyLocator.ts#L28

### 👍 Expected behavior

It should work the same it does in Node 16.x

### 👎 Actual Behavior with Screenshots

Fetch fails for the item inside the service request:
Screenshot 2023-09-13 at 3 00 28 PM
Screenshot 2023-09-13 at 3 01 57 PM

### 👟 Reproduction steps

1. Grab master branch
2. Apply the dice-roller manifest in your local cluster from `plugins/kubernetes-backend/examples/dice-roller/dice-roller-manifests.yaml` (might need to fix the http path prefix if using latest k8s version).
3. Use Node 18.x
4. Run `kubectl proxy`
5. `yarn dev`
6. Go to the catalog for the dice-roller service, in the k8s tab the error should show.

### 📃 Provide the context for the Bug.

_No response_

### 🖥️ Your Environment

OS:   Darwin 22.6.0 - darwin/x64
node: v16.20.2
yarn: 3.2.3
cli:  0.22.13-next.3 (local)
backstage:  N/A

<details><summary>Dependencies:</summary>
<p>
  @backstage/app-defaults                                            0.0.0-use.local
  @backstage/backend-app-api                                         0.0.0-use.local
  @backstage/backend-common                                          0.0.0-use.local
  @backstage/backend-defaults                                        0.0.0-use.local
  @backstage/backend-dev-utils                                       0.0.0-use.local
  @backstage/backend-openapi-utils                                   0.0.0-use.local
  @backstage/backend-plugin-api                                      0.0.0-use.local
  @backstage/backend-plugin-manager                                  0.0.0-use.local
  @backstage/backend-tasks                                           0.0.0-use.local
  @backstage/backend-test-utils                                      0.0.0-use.local
  @backstage/catalog-client                                          1.4.3, 0.0.0-use.local
  @backstage/catalog-model                                           1.4.1, 0.0.0-use.local
  @backstage/cli-common                                              0.0.0-use.local
  @backstage/cli-node                                                0.0.0-use.local
  @backstage/cli                                                     0.0.0-use.local
  @backstage/codemods                                                0.0.0-use.local
  @backstage/config-loader                                           0.0.0-use.local
  @backstage/config                                                  1.0.8, 0.0.0-use.local
  @backstage/core-app-api                                            0.0.0-use.local
  @backstage/core-components                                         0.12.5, 0.13.4, 0.0.0-use.local
  @backstage/core-plugin-api                                         1.5.3, 0.0.0-use.local
  @backstage/create-app                                              0.0.0-use.local
  @backstage/dev-utils                                               0.0.0-use.local
  @backstage/errors                                                  1.2.1, 0.0.0-use.local
  @backstage/eslint-plugin                                           0.0.0-use.local
  @backstage/frontend-app-api                                        0.0.0-use.local
  @backstage/frontend-plugin-api                                     0.0.0-use.local
  @backstage/integration-aws-node                                    0.0.0-use.local
  @backstage/integration-react                                       1.1.18, 0.0.0-use.local
  @backstage/integration                                             1.6.2, 0.0.0-use.local
  @backstage/plugin-adr-backend                                      0.0.0-use.local
  @backstage/plugin-adr-common                                       0.0.0-use.local
  @backstage/plugin-adr                                              0.0.0-use.local
  @backstage/plugin-airbrake-backend                                 0.0.0-use.local
  @backstage/plugin-airbrake                                         0.0.0-use.local
  @backstage/plugin-allure                                           0.0.0-use.local
  @backstage/plugin-analytics-module-ga4                             0.0.0-use.local
  @backstage/plugin-analytics-module-ga                              0.0.0-use.local
  @backstage/plugin-analytics-module-newrelic-browser                0.0.0-use.local
  @backstage/plugin-apache-airflow                                   0.0.0-use.local
  @backstage/plugin-api-docs-module-protoc-gen-doc                   0.0.0-use.local
  @backstage/plugin-api-docs                                         0.0.0-use.local
  @backstage/plugin-apollo-explorer                                  0.0.0-use.local
  @backstage/plugin-app-backend                                      0.0.0-use.local
  @backstage/plugin-app-node                                         0.0.0-use.local
  @backstage/plugin-auth-backend-module-gcp-iap-provider             0.0.0-use.local
  @backstage/plugin-auth-backend-module-github-provider              0.0.0-use.local
  @backstage/plugin-auth-backend-module-gitlab-provider              0.0.0-use.local
  @backstage/plugin-auth-backend-module-google-provider              0.0.0-use.local
  @backstage/plugin-auth-backend-module-oauth2-provider              0.0.0-use.local
  @backstage/plugin-auth-backend                                     0.0.0-use.local
  @backstage/plugin-auth-node                                        0.0.0-use.local
  @backstage/plugin-azure-devops-backend                             0.0.0-use.local
  @backstage/plugin-azure-devops-common                              0.0.0-use.local
  @backstage/plugin-azure-devops                                     0.0.0-use.local
  @backstage/plugin-azure-sites-backend                              0.0.0-use.local
  @backstage/plugin-azure-sites-common                               0.0.0-use.local
  @backstage/plugin-azure-sites                                      0.0.0-use.local
  @backstage/plugin-badges-backend                                   0.0.0-use.local
  @backstage/plugin-badges                                           0.0.0-use.local
  @backstage/plugin-bazaar-backend                                   0.0.0-use.local
  @backstage/plugin-bazaar                                           0.0.0-use.local
  @backstage/plugin-bitbucket-cloud-common                           0.0.0-use.local
  @backstage/plugin-bitrise                                          0.0.0-use.local
  @backstage/plugin-catalog-backend-module-aws                       0.0.0-use.local
  @backstage/plugin-catalog-backend-module-azure                     0.0.0-use.local
  @backstage/plugin-catalog-backend-module-bitbucket-cloud           0.0.0-use.local
  @backstage/plugin-catalog-backend-module-bitbucket-server          0.0.0-use.local
  @backstage/plugin-catalog-backend-module-bitbucket                 0.0.0-use.local
  @backstage/plugin-catalog-backend-module-gcp                       0.0.0-use.local
  @backstage/plugin-catalog-backend-module-gerrit                    0.0.0-use.local
  @backstage/plugin-catalog-backend-module-github                    0.0.0-use.local
  @backstage/plugin-catalog-backend-module-gitlab                    0.0.0-use.local
  @backstage/plugin-catalog-backend-module-incremental-ingestion     0.0.0-use.local
  @backstage/plugin-catalog-backend-module-ldap                      0.0.0-use.local
  @backstage/plugin-catalog-backend-module-msgraph                   0.0.0-use.local
  @backstage/plugin-catalog-backend-module-openapi                   0.0.0-use.local
  @backstage/plugin-catalog-backend-module-puppetdb                  0.0.0-use.local
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model   0.0.0-use.local
  @backstage/plugin-catalog-backend-module-unprocessed               0.0.0-use.local
  @backstage/plugin-catalog-backend                                  0.0.0-use.local
  @backstage/plugin-catalog-common                                   1.0.15, 0.0.0-use.local
  @backstage/plugin-catalog-graph                                    0.0.0-use.local
  @backstage/plugin-catalog-graphql                                  0.0.0-use.local
  @backstage/plugin-catalog-import                                   0.0.0-use.local
  @backstage/plugin-catalog-node                                     0.0.0-use.local
  @backstage/plugin-catalog-react                                    1.8.3, 0.0.0-use.local
  @backstage/plugin-catalog-unprocessed-entities                     0.0.0-use.local
  @backstage/plugin-catalog                                          0.0.0-use.local
  @backstage/plugin-cicd-statistics-module-gitlab                    0.0.0-use.local
  @backstage/plugin-cicd-statistics                                  0.0.0-use.local
  @backstage/plugin-circleci                                         0.0.0-use.local
  @backstage/plugin-cloudbuild                                       0.0.0-use.local
  @backstage/plugin-code-climate                                     0.0.0-use.local
  @backstage/plugin-code-coverage-backend                            0.0.0-use.local
  @backstage/plugin-code-coverage                                    0.0.0-use.local
  @backstage/plugin-codescene                                        0.0.0-use.local
  @backstage/plugin-config-schema                                    0.0.0-use.local
  @backstage/plugin-cost-insights-common                             0.0.0-use.local
  @backstage/plugin-cost-insights                                    0.0.0-use.local
  @backstage/plugin-devtools-backend                                 0.0.0-use.local
  @backstage/plugin-devtools-common                                  0.0.0-use.local
  @backstage/plugin-devtools                                         0.0.0-use.local
  @backstage/plugin-dynatrace                                        0.0.0-use.local
  @backstage/plugin-entity-feedback-backend                          0.0.0-use.local
  @backstage/plugin-entity-feedback-common                           0.0.0-use.local
  @backstage/plugin-entity-feedback                                  0.0.0-use.local
  @backstage/plugin-entity-validation                                0.0.0-use.local
  @backstage/plugin-events-backend-module-aws-sqs                    0.0.0-use.local
  @backstage/plugin-events-backend-module-azure                      0.0.0-use.local
  @backstage/plugin-events-backend-module-bitbucket-cloud            0.0.0-use.local
  @backstage/plugin-events-backend-module-gerrit                     0.0.0-use.local
  @backstage/plugin-events-backend-module-github                     0.0.0-use.local
  @backstage/plugin-events-backend-module-gitlab                     0.0.0-use.local
  @backstage/plugin-events-backend-test-utils                        0.0.0-use.local
  @backstage/plugin-events-backend                                   0.0.0-use.local
  @backstage/plugin-events-node                                      0.0.0-use.local
  @backstage/plugin-explore-backend                                  0.0.0-use.local
  @backstage/plugin-explore-common                                   0.0.0-use.local
  @backstage/plugin-explore-react                                    0.0.0-use.local
  @backstage/plugin-explore                                          0.0.0-use.local
  @backstage/plugin-firehydrant                                      0.0.0-use.local
  @backstage/plugin-fossa                                            0.0.0-use.local
  @backstage/plugin-gcalendar                                        0.0.0-use.local
  @backstage/plugin-gcp-projects                                     0.0.0-use.local
  @backstage/plugin-git-release-manager                              0.0.0-use.local
  @backstage/plugin-github-actions                                   0.0.0-use.local
  @backstage/plugin-github-deployments                               0.0.0-use.local
  @backstage/plugin-github-issues                                    0.0.0-use.local
  @backstage/plugin-github-pull-requests-board                       0.0.0-use.local
  @backstage/plugin-gitops-profiles                                  0.0.0-use.local
  @backstage/plugin-gocd                                             0.0.0-use.local
  @backstage/plugin-graphiql                                         0.0.0-use.local
  @backstage/plugin-graphql-backend                                  0.0.0-use.local
  @backstage/plugin-graphql-voyager                                  0.0.0-use.local
  @backstage/plugin-home-react                                       0.1.2, 0.0.0-use.local
  @backstage/plugin-home                                             0.5.7, 0.0.0-use.local
  @backstage/plugin-ilert                                            0.0.0-use.local
  @backstage/plugin-jenkins-backend                                  0.0.0-use.local
  @backstage/plugin-jenkins-common                                   0.0.0-use.local
  @backstage/plugin-jenkins                                          0.0.0-use.local
  @backstage/plugin-kafka-backend                                    0.0.0-use.local
  @backstage/plugin-kafka                                            0.0.0-use.local
  @backstage/plugin-kubernetes-backend                               0.0.0-use.local
  @backstage/plugin-kubernetes-common                                0.0.0-use.local
  @backstage/plugin-kubernetes                                       0.0.0-use.local
  @backstage/plugin-lighthouse-backend                               0.0.0-use.local
  @backstage/plugin-lighthouse-common                                0.0.0-use.local
  @backstage/plugin-lighthouse                                       0.0.0-use.local
  @backstage/plugin-linguist-backend                                 0.0.0-use.local
  @backstage/plugin-linguist-common                                  0.0.0-use.local
  @backstage/plugin-linguist                                         0.0.0-use.local
  @backstage/plugin-microsoft-calendar                               0.0.0-use.local
  @backstage/plugin-newrelic-dashboard                               0.0.0-use.local
  @backstage/plugin-newrelic                                         0.0.0-use.local
  @backstage/plugin-nomad-backend                                    0.0.0-use.local
  @backstage/plugin-nomad                                            0.0.0-use.local
  @backstage/plugin-octopus-deploy                                   0.0.0-use.local
  @backstage/plugin-opencost                                         0.0.0-use.local
  @backstage/plugin-org-react                                        0.0.0-use.local
  @backstage/plugin-org                                              0.0.0-use.local
  @backstage/plugin-pagerduty                                        0.0.0-use.local
  @backstage/plugin-periskop-backend                                 0.0.0-use.local
  @backstage/plugin-periskop                                         0.0.0-use.local
  @backstage/plugin-permission-backend-module-allow-all-policy       0.0.0-use.local
  @backstage/plugin-permission-backend                               0.0.0-use.local
  @backstage/plugin-permission-common                                0.7.7, 0.0.0-use.local
  @backstage/plugin-permission-node                                  0.0.0-use.local
  @backstage/plugin-permission-react                                 0.4.14, 0.0.0-use.local
  @backstage/plugin-playlist-backend                                 0.0.0-use.local
  @backstage/plugin-playlist-common                                  0.0.0-use.local
  @backstage/plugin-playlist                                         0.0.0-use.local
  @backstage/plugin-proxy-backend                                    0.0.0-use.local
  @backstage/plugin-puppetdb                                         0.0.0-use.local
  @backstage/plugin-rollbar-backend                                  0.0.0-use.local
  @backstage/plugin-rollbar                                          0.0.0-use.local
  @backstage/plugin-scaffolder-backend-module-confluence-to-markdown 0.0.0-use.local
  @backstage/plugin-scaffolder-backend-module-cookiecutter           0.0.0-use.local
  @backstage/plugin-scaffolder-backend-module-gitlab                 0.0.0-use.local
  @backstage/plugin-scaffolder-backend-module-rails                  0.0.0-use.local
  @backstage/plugin-scaffolder-backend-module-sentry                 0.0.0-use.local
  @backstage/plugin-scaffolder-backend-module-yeoman                 0.0.0-use.local
  @backstage/plugin-scaffolder-backend                               0.0.0-use.local
  @backstage/plugin-scaffolder-common                                0.0.0-use.local
  @backstage/plugin-scaffolder-node                                  0.0.0-use.local
  @backstage/plugin-scaffolder-react                                 0.0.0-use.local
  @backstage/plugin-scaffolder                                       0.0.0-use.local
  @backstage/plugin-search-backend-module-catalog                    0.0.0-use.local
  @backstage/plugin-search-backend-module-elasticsearch              0.0.0-use.local
  @backstage/plugin-search-backend-module-explore                    0.0.0-use.local
  @backstage/plugin-search-backend-module-pg                         0.0.0-use.local
  @backstage/plugin-search-backend-module-techdocs                   0.0.0-use.local
  @backstage/plugin-search-backend-node                              0.0.0-use.local
  @backstage/plugin-search-backend                                   0.0.0-use.local
  @backstage/plugin-search-common                                    1.2.5, 0.0.0-use.local
  @backstage/plugin-search-react                                     0.0.0-use.local
  @backstage/plugin-search                                           0.0.0-use.local
  @backstage/plugin-sentry                                           0.0.0-use.local
  @backstage/plugin-shortcuts                                        0.0.0-use.local
  @backstage/plugin-sonarqube-backend                                0.0.0-use.local
  @backstage/plugin-sonarqube-react                                  0.0.0-use.local
  @backstage/plugin-sonarqube                                        0.0.0-use.local
  @backstage/plugin-splunk-on-call                                   0.0.0-use.local
  @backstage/plugin-stack-overflow-backend                           0.0.0-use.local
  @backstage/plugin-stack-overflow                                   0.0.0-use.local
  @backstage/plugin-stackstorm                                       0.0.0-use.local
  @backstage/plugin-tech-insights-backend-module-jsonfc              0.0.0-use.local
  @backstage/plugin-tech-insights-backend                            0.0.0-use.local
  @backstage/plugin-tech-insights-common                             0.0.0-use.local
  @backstage/plugin-tech-insights-node                               0.0.0-use.local
  @backstage/plugin-tech-insights                                    0.0.0-use.local
  @backstage/plugin-tech-radar                                       0.0.0-use.local
  @backstage/plugin-techdocs-addons-test-utils                       0.0.0-use.local
  @backstage/plugin-techdocs-backend                                 0.0.0-use.local
  @backstage/plugin-techdocs-module-addons-contrib                   0.0.0-use.local
  @backstage/plugin-techdocs-node                                    0.0.0-use.local
  @backstage/plugin-techdocs-react                                   0.0.0-use.local
  @backstage/plugin-techdocs                                         0.0.0-use.local
  @backstage/plugin-todo-backend                                     0.0.0-use.local
  @backstage/plugin-todo                                             0.0.0-use.local
  @backstage/plugin-user-settings-backend                            0.0.0-use.local
  @backstage/plugin-user-settings                                    0.0.0-use.local
  @backstage/plugin-vault-backend                                    0.0.0-use.local
  @backstage/plugin-vault                                            0.0.0-use.local
  @backstage/plugin-xcmetrics                                        0.0.0-use.local
  @backstage/release-manifests                                       0.0.0-use.local
  @backstage/repo-tools                                              0.0.0-use.local
  @backstage/test-utils                                              0.0.0-use.local
  @backstage/theme                                                   0.2.19, 0.4.1, 0.0.0-use.local
  @backstage/types                                                   1.1.0, 0.0.0-use.local
  @backstage/version-bridge                                          1.0.4, 0.0.0-use.local

</p>
</details> 

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

### PR DESCRIPTION
**Pull request by luchillo17**
## Hey, I just made a Pull Request!

The gist of it is, there was a breaking change in NodeJS 17.x that changed the default DNS IP resolving behavior to IPv6, to fix it we need to tell the Node DNS to resolve to IPv4 first.

Fixes #19931

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/backend/src/index.ts
Comment (jamieklassen) [inline packages/backend/src/index.ts:78]: It [looks](https://nodejs.org/api/dns.html#dnssetdefaultresultorderorder) like this setting is global (or at least per-thread) -- if folks like the approach (I leave it to maintainers with broader perspective than mine), then it might make sense to put this setting in the [`create-app` template](https://github.com/backstage/backstage/blob/master/packages/create-app/templates/default-app/packages/backend/src/index.ts) as well.

Comment (luchillo17) [reply packages/backend/src/index.ts:78]: Does the template not use this package? I would assume just using the backend package would trigger this line...

Comment (jamieklassen) [reply packages/backend/src/index.ts:78]: They are separate -- for example, the sample app has many plugins installed for testing, but an freshly-scaffolded app is more "blank". Often changes like this need to be duplicated in the templates, but as I said, this whole workflow is about local development so a fix like this might not need to be part of the end-user templates.

Comment (freben) [reply packages/backend/src/index.ts:78]: Indeed `packages/backend` is just a test backend that we use ourselves for local testing, and has nothing at all to do with backends that end users see.

Comment (luchillo17) [reply packages/backend/src/index.ts:78]: How likely is it for production deployments to use `kubectl proxy`? is it ok to assume this is only used in testing environments & product demos?

Comment (freben) [reply packages/backend/src/index.ts:78]: No those aren't released as packages, so they typically don't get any changesets. The only people potentially affected are those that do local development _on the oss repo_ specifically, and only if they run the example backend in the first place instead of doing local testing of their specific plugin.

**Comment by luchillo17** [comment]
Ping avoid stale.

**Comment by luchillo17** [comment]
Ping no stale.

**Review by jamieklassen — APPROVED** [review]

File: plugins/kubernetes-backend/src/cluster-locator/LocalKubectlProxyLocator.ts
Comment (Rugvip) [inline plugins/kubernetes-backend/src/cluster-locator/LocalKubectlProxyLocator.ts:29]: I'm not entirely sure I follow what is triggering the actual issue here. Is it that we don't know whether the kube proxy is listening to IPv4 or IPv6, and there are situations where only one of them will be available?

I wonder if we could do the resolution upfront here in that case? either if we could do a lookup for example with https://nodejs.org/api/dns.html#dnslookuphostname-options-callback, or if we just try to connect over both 4 and 6 and pick one that succeeds?

**Review by Rugvip — CHANGES_REQUESTED** [review]
I think it'd be good to find a solution that helps everyone rather than only gets this working in the main repo

**Comment by freben** [comment]
@luchillo17 Getting back to this now. So, this PR really doesn't help anyone right, besides those who happen to contribute to the main Backstage repo and happen to have an org for whose dev machines this change is good rather than bad.

Makes me wonder what the big picture really should be. I'm not sure that adding this in the create-app template for example will be safe, since messing with the resolution order probably has unintended consequences sometimes. What do you think? Happy for input on this

File: plugins/kubernetes-backend/src/cluster-locator/LocalKubectlProxyLocator.ts
Comment (luchillo17) [reply plugins/kubernetes-backend/src/cluster-locator/LocalKubectlProxyLocator.ts:29]: This is doing nothing to fix the issue, just fixing what is very likely to have been a typo.

**Comment by luchillo17** [comment]
@freben It does help people, people who use Node 18.x (maybe even higher versions, haven't tested with other versions, current LTS is 20.10.x) that contribute to backstage are likely to run into this, I know I did...

File: plugins/kubernetes-backend/src/cluster-locator/LocalKubectlProxyLocator.ts
Comment (Rugvip) [reply plugins/kubernetes-backend/src/cluster-locator/LocalKubectlProxyLocator.ts:29]: Yeah I wasn't talking about this exact change, but rather the issue you're aiming to fix. I'm suggesting we implement the logic I proposed in this part of the code

**Comment by luchillo17** [comment]
@Rugvip Let's continue the thread of the review here, basically based on the conversation on #19931 and this PR, from Node 17.x up the default DNS resolve order was changed to IPv6 first, which doesn't like the `localhost` host, the NodeJS thread I found this information offered 2 solutions:

1. We change `localhost` for `127.0.0.1` loopback address, which I was told would break some currently supported dev environments like WSL2 and VMS.
2. We purposely tell NodeJS to revert back to old behavior to resolve DNS by IPv4 first, which is the main change of this PR.

**Comment by Rugvip** [comment]
@luchillo17 can we do a lookup upfront and substitute `localhost` for whichever of `127.0.0.1` and `::1` that we can successfully connect to? I'm assuming still that this is something people will run into if they try to use the kubernetes plugin in dev?

**Comment by luchillo17** [comment]
Dunno, and at the moment I can't sign my commits, as of today I don't work for VMware anymore 🥲

**Pull request closed without merging**
